trieste/acquisition/function.py

```diff
@@ -405,3 +405,93 @@ def probability_of_feasibility(
+        series={Proceedings of Machine Learning Research},
+        month={22--24 Jun},
+        publisher={PMLR},
+        url={http://proceedings.mlr.press/v32/gardner14.html},
+    }
+
+    """
+
+    _PENALIZATION_FACTOR = 1.1
+
+    def __init__(
+        self,
+        objective_tag: str,
+        constraint_builder: AcquisitionFunctionBuilder,
+        min_feasibility_probability: Union[float, tf.Tensor] = 0.5,
+    ):
+        """
+        :param objective_tag: The tag for the objective data and model.
+        :param constraint_builder: The builder for the constraint function.
+        :param min_feasibility_probability: The minimum probability of feasibility for a
+            "best point" to be considered feasible.
+        :raise ValueError (or InvalidArgumentError): If ``min_feasibility_probability`` is not a
+            scalar in the unit interval :math:`[0, 1]`.
+        """
+        tf.debugging.assert_scalar(min_feasibility_probability)
+
+        if not 0 <= min_feasibility_probability <= 1:
+            raise ValueError(
+                f"Minimum feasibility probability must be between 0 and 1 inclusive,"
+                f" got {min_feasibility_probability}"
+            )
+
+        self._objective_tag = objective_tag
+        self._constraint_builder = constraint_builder
+        self._min_feasibility_probability = min_feasibility_probability
+
+    def prepare_acquisition_function(
+        self, datasets: Mapping[str, Dataset], models: Mapping[str, ModelInterface]
+    ) -> AcquisitionFunction:
+        """
+        :param datasets: The data from the observer.
+        :param models: The models over each dataset in ``datasets``.
+        :return: The expected constrained improvement acquisition function.
+        :raise KeyError: If `objective_tag` is not found in ``datasets`` and ``models``.
+        :raise ValueError: If the objective data is empty, or no feasible points are found in
+            ``datasets``.
+        """
+        objective_model = models[self._objective_tag]
+        objective_dataset = datasets[self._objective_tag]
+
+        if len(objective_dataset) == 0:
+            raise ValueError(
+                "Expected improvement is defined with respect to existing points in the objective"
+                " data, but the objective data is empty."
+            )
+
+        constraint_fn = self._constraint_builder.prepare_acquisition_function(datasets, models)
+        pof = constraint_fn(objective_dataset.query_points)
+        is_not_feasible = pof < self._min_feasibility_probability
+
+        if tf.reduce_all(is_not_feasible):
+            return constraint_fn
+
+        mean, _ = objective_model.predict(objective_dataset.query_points)
+        range_of_mean = tf.reduce_max(mean) - tf.reduce_min(mean)
+        penalization = self._PENALIZATION_FACTOR * range_of_mean * tf.cast(
+            is_not_feasible, mean.dtype
+        )
+        eta = tf.reduce_min(mean + penalization, axis=0)
```

Comment:
this is unnecessarily complex if we return `constraint_fn` when no feasible point has been identified. We don't need `range_of_mean`, and PENALIZATION_FACTOR can be set to `+inf`.

Reply:
if we're going to do that, we can just mask out any unfeasible points, I think

Reply:
like
```python
        is_feasible = pof >= self._min_feasibility_probability

        if not tf.reduce_any(is_feasible):
            return constraint_fn

        mean, _ = objective_model.predict(objective_dataset.query_points)
        eta = tf.reduce_min(tf.boolean_mask(mean, is_feasible), axis=0)
```

Reply:
yes, that sounds perfect